### PR TITLE
Use `go-git` functionalities for common git operations

### PR DESCRIPTION
**Pull request by A-Hilaly**
Part of https://github.com/aws-controllers-k8s/community/issues/835

This patch updates all the helper functions to use `go-git` library
instead of having to call `os/exec.Command`/`os/exec.CombinedOutput` in
order to clone repository, fetch tags... These changes brings more
dependencies to our `go.mod` but in exchange it simplifies the way we
interact with git repositories and improves errors handling and
readability.

Future changes will also benefit from this patch, especially
conversation webhook generator that will need to checkout multiple
version tags.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by A-Hilaly** [comment]
Needs `ATTRIBUTIONS.md` update
/hold

File: cmd/ack-generate/command/common.go [hunk 1]
```diff
@@ -64,67 +66,93 @@ func isDirWriteable(fp string) bool {
...
+		return fmt.Errorf("cannot checkout tag: %v", err)
 	}
-	return clonePath, nil
+
+	return err
+}
+
+// correctSemanticVersion takes a semver string and tries to append the 'v'
+// prefix if it's missing.
+func correctSemanticVersion(s string) string {
```
Comment (RedbackThomson) [inline cmd/ack-generate/command/common.go:123]: nit: Method name is a bit ambiguous. Perhaps "ensureSemverPrefix"?

File: cmd/ack-generate/command/common.go [hunk 2]
```diff
@@ -64,67 +66,93 @@ func isDirWriteable(fp string) bool {
...
-		return clonePath, cmd.Run()
+
+	// get sdkVersion and correct it (adds missing 'v')
+	sdkVersion, err := getSDKVersion("")
+	if err != nil {
+		return err
+	}
+	sdkVersion = correctSemanticVersion(sdkVersion)
+
+	// Now checkout the local repository.
```
Comment (RedbackThomson) [inline cmd/ack-generate/command/common.go:107]: Comment doesn't match? `LoadRepository` doesn't `git checkout` - just loads into memory?

File: cmd/ack-generate/command/common.go [hunk 3]
```diff
@@ -64,67 +66,93 @@ func isDirWriteable(fp string) bool {
...
+	return fmt.Sprintf("v%s", s)
 }
 
 // getSDKVersion returns the github.com/aws/aws-sdk-go version to use. It
-// first tries to get return version from the --aws-sdk-go-version flag, then
-// look for the service controller and local go.mod files.
-func getSDKVersion() (string, error) {
+// first tries to get the version from the --aws-sdk-go-version flag, then
+// from the ack-generate-metadata.yaml and finally look for the service
+// go.mod controller.
```
Comment (RedbackThomson) [inline cmd/ack-generate/command/common.go:133]: 👌

File: pkg/util/git.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		return nil
+	}
+	return err
+}
+
+// GetRepositoryTagRef returns the git reference (commit hash) of a given tag.
+// NOTE: It is not possible to checkout a tag without knowing it's reference.
+//
+// Calling this function is equivalent to executing `git rev-list -n 1 $tagName`
+func GetRepositoryTagRef(repo *git.Repository, tagName string) (*plumbing.Reference, error) {
```
Comment (RedbackThomson) [inline pkg/util/git.go:74]: nit: Make into a private method?

File: pkg/util/git.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	"errors"
+	"fmt"
+	"io"
+
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+// LoadRepository loads a repository from the local file system.
+// TODO(a-hilaly): load repository into a memory filesystem (needs go1.16
```
Comment (RedbackThomson) [inline pkg/util/git.go:27]: How large would this be in memory? I know the `aws-sdk-go` has a large commit history and many, many tags.

File: cmd/ack-generate/command/common.go (hunk 2 above)
Comment (A-Hilaly) [reply cmd/ack-generate/command/common.go:107]: Yeah it was more for the rest of the function body (`LoadRepository` + `CheckoutRepository`). Moving the comment

File: pkg/util/git.go (hunk 4 above)
Comment (A-Hilaly) [reply pkg/util/git.go:74]: ++

File: pkg/util/git.go (hunk 5 above)
Comment (A-Hilaly) [reply pkg/util/git.go:27]:
```shell
du -sh ~/.cache/aws-controllers-k8s/src/aws-sdk-go 
352M	/Users/hilalymh/.cache/aws-controllers-k8s/src/aws-sdk-go
```

I guess it's gonna be ~350M - I will do some profile on `ack-generate` binary before implementing it

Comment (A-Hilaly) [reply pkg/util/git.go:27]: @RedbackThomson Is it a bad a idea to load such a big repository in memory?

**Comment by vijtrip2** [comment]
/approve

I am sure you tested this locally for some controller. Good to add the output in PR since there are no new unit tests.

**Comment by A-Hilaly** [comment]
Adding unit tests
/hold

**Comment by A-Hilaly** [comment]
Looks like it's very complex to unit test these util functions, and it's almost equivalent of testing the go-git library. @vijtrip2 and I agreed on leaving the PR as is and just run some local tests.

On my local machine
```bash
# First run (no local aws-sdk-go cache)
time ACK_GENERATE_CACHE_DIR=/tmp/gitest AWS_SDK_GO_VERSION=v1.38.60 make build-ack-generate build-controller SERVICE=ecr
building ack-generate ... ok.
Copying common custom resource definitions into ecr
Building Kubernetes API objects for ecr
Generating deepcopy code for ecr
Generating custom resource definitions for ecr
Building service controller for ecr
Generating RBAC manifests for ecr
Running gofmt against generated code for ecr
ACK_GENERATE_CACHE_DIR=/tmp/gitest AWS_SDK_GO_VERSION=v1.38.60 make     54,08s user 3,20s system 50% cpu 1:53,98 total

# Second run
time ACK_GENERATE_CACHE_DIR=/tmp/gitest AWS_SDK_GO_VERSION=v1.38.60 make build-ack-generate build-controller SERVICE=ecr
building ack-generate ... ok.
Copying common custom resource definitions into ecr
Building Kubernetes API objects for ecr
Generating deepcopy code for ecr
Generating custom resource definitions for ecr
Building service controller for ecr
Generating RBAC manifests for ecr
Running gofmt against generated code for ecr
ACK_GENERATE_CACHE_DIR=/tmp/gitest AWS_SDK_GO_VERSION=v1.38.60 make     8,04s user 1,27s system 176% cpu 5,264 total
```

/unhold

**Review by jaypipes — APPROVED** [review]
Nice and clean, @A-Hilaly. Good job! :)

**Pull request merged**